lib/SILGen/SILGenExpr.cpp

```diff
@@ -2351,9 +2350,11 @@ static bool mayLieAboutNonOptionalReturn(Expr *expr) {
     return mayLieAboutNonOptionalReturn(declRef->getDecl());
   }
 
-  // An application, which we look through to get the function we're calling.
+  // An application that produces a reference type, which we look through to
+  // get the function we're calling.
   if (auto apply = dyn_cast<ApplyExpr>(expr)) {
-    return mayLieAboutNonOptionalReturn(apply->getFn());
+    return apply->getType()->hasReferenceSemantics() &&
```

Comment:
@jrose-apple It isn't clear to me how `mayLieAboutNonOptionalReturn` currently handles `UnsafePointer`s. AFAICT it only checks `hasReferenceSemantics`, which is only true for class, metatype, and function types.